src/LanguageServer.Engine/Documents/MasterProjectDocument.cs

```diff
@@ -67,15 +68,21 @@ protected override void Dispose(bool disposing)
         /// <summary>
         ///     Add a sub-project.
         /// </summary>
-        /// <param name="subProjectDocument">
+        /// <param name="documentUri">
         ///     The sub-project.
         /// </param>
-        public void AddSubProject(SubProjectDocument subProjectDocument)
+        /// <param name="createSubProjectDocument">
+        ///     A factory delegate to create the <see cref="SubProjectDocument"/> if it does not already exist.
+        /// </param>
+        public SubProjectDocument GetOrAddSubProject(Uri documentUri, Func<SubProjectDocument> createSubProjectDocument)
```

Comment:
Name for the second parameter should probably be like `subProjectDocumentFactory` or something along those lines. `CreateSubProjectDocument` is a good name for a method, but not a parametrer

Reply:
Yeah I’ve never fully agreed with that idea (delegates being a special case, e.g. calling one “configure”) but I can see we’d use factory elsewhere so will change it to match.

Reply:
Having checked the rest of the project's code, the only other place we have a similar factory delegate it follows this same pattern:

https://github.com/tintoy/msbuild-project-tools-server/blob/0c28300c30f28950c647501200f4dc86ad6aa17f/src/LanguageServer.SemanticModel.MSBuild/MSBuildExpressions/Parsers.cs#L768

(so I'm going to leave it as-is for now)